src/middlewared/middlewared/plugins/account.py

```diff
@@ -27,50 +81,477 @@ class UserService(CRUDService):
+            if gm['id'] not in groups:
+                await self.middleware.call('datastore.delete', 'account.bsdgroupmembership', gm['id'])
+            else:
+                existing_ids.add(gm['id'])
+
+        for _id in groups - existing_ids:
+            group = await self.middleware.call('datastore.query', 'account.bsdgroups', [('id', '=', _id)], {'prefix': 'bsdgrp_'})
+            if not group:
+                raise CallError(f'Group {_id} not found', errno.ENOENT)
+            await self.middleware.call(
+                'datastore.insert',
+                'account.bsdgroupmembership',
+                {'group': _id, 'user': pk},
+                {'prefix': 'bsdgrpmember_'}
+            )
+
+    async def __update_sshpubkey(self, user, group):
+        if 'sshpubkey' not in user:
+            return
+        keysfile = f'{user["home"]}/.ssh/authorized_keys'
+        pubkey = user.pop('sshpubkey')
+        if pubkey is None:
+            if os.path.exists(keysfile):
+                try:
+                    os.unlink(keysfile)
+                except OSError:
                     pass
-                pubkey = pubkey.strip() + '\n'
-                if pubkey != oldpubkey:
-                    sshpath = f'{user["home"]}/.ssh'
-                    saved_umask = os.umask(0o77)
-                    if not os.path.isdir(sshpath):
-                        os.makedirs(sshpath)
-                    if not os.path.isdir(sshpath):
-                        raise CallError(f'{sshpath} is not a directory')
-                    if pubkey == '' and os.path.exists(keysfile):
-                        os.unlink(keysfile)
-                    else:
-                        with open(keysfile, 'w') as f:
-                            f.write(pubkey)
-                        await run('chown', '-R', f'{user["username"]}:{user["group"]["bsdgrp_group"]}', sshpath, check=False)
-                    os.umask(saved_umask)
+        else:
+            oldpubkey = ''
+            try:
+                with open(keysfile, 'r') as f:
+                    oldpubkey = f.read()
+            except Exception:
+                pass
+            pubkey = pubkey.strip() + '\n'
+            if pubkey != oldpubkey:
+                sshpath = f'{user["home"]}/.ssh'
+                saved_umask = os.umask(0o77)
+                if not os.path.isdir(sshpath):
+                    os.makedirs(sshpath)
+                if not os.path.isdir(sshpath):
+                    raise CallError(f'{sshpath} is not a directory')
+                if pubkey == '' and os.path.exists(keysfile):
+                    os.unlink(keysfile)
+                else:
+                    with open(keysfile, 'w') as f:
+                        f.write(pubkey)
+                    await run('chown', '-R', f'{user["username"]}:{group}', sshpath, check=False)
+                os.umask(saved_umask)
```

Comment:
Why umask is needed here? We may just run chmod for `.ssh` directory and keyfile